File: 3. Behavioral Design Patterns/Command/Javascript/CommandTemplate.js
// Command Pattern – JavaScript Template

class Command { execute(){} undo(){} }

class Receiver { action(){ /* do something */ } undoAction(){ /* undo something */ } }

class ConcreteCommand extends Command {
  constructor(r){ super(); this.r = r }
  execute(){ this.r.action() }
  undo(){ this.r.undoAction() }
}

class Invoker { constructor(){ this.cmd = null } set(c){ this.cmd = c } run(){ if(this.cmd) this.cmd.execute() } undo(){ if(this.cmd) this.cmd.undo() } }

;(function main(){ const inv = new Invoker(); inv.set(new ConcreteCommand(new Receiver())); inv.run(); inv.undo() })()

File: 3. Behavioral Design Patterns/Command/Python/CommandTemplate.py
# Command Pattern – Python Template

from abc import ABC, abstractmethod

class Command(ABC):
    @abstractmethod
    def execute(self):
        pass
    
    @abstractmethod
    def undo(self):
        pass

class Receiver:
    def action(self):
        return "Receiver action"
    
    def undo_action(self):
        return "Receiver undo action"

class ConcreteCommand(Command):
    def __init__(self, receiver):
        self.receiver = receiver
    
    def execute(self):
        return self.receiver.action()
    
    def undo(self):
        return self.receiver.undo_action()

class Invoker:
    def __init__(self):
        self.command = None
    
    def set_command(self, command):
        self.command = command
    
    def run(self):
        if self.command:
            return self.command.execute()
    
    def undo(self):
        if self.command:
            return self.command.undo()

def main():
    receiver = Receiver()
    command = ConcreteCommand(receiver)
    invoker = Invoker()
    invoker.set_command(command)
    print(invoker.run())
    print(invoker.undo())

if __name__ == "__main__":
    main()
